### PR TITLE
Simplify internal Buffer allocations

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
There are two concerns addressed here:
1. The use of UntetheredMemory outside of the pooling allocator was more complexity than was strictly necessary, and also created an awkward API dance where the memory and its drop had to be combined again.
2. MemoryManager instances may want to create Buffers and their baseline Drop instance together, yet the allocator need to adapt or decorate the drop.

Modification:
Change the ensureWritable implementations to no longer rely on UntetheredMemory, but instead allocate a new Buffer instance and cannibalise its insides.
Move the UntetheredMemory interface into the buffer/api/pool package.
Remove the MemoryManager.drop method and change the MemoryManager.allocateShared to take an adaptor function.
The memory managers can then create per-Buffer drop instances if they so desire.
Simplify how the PooledBufferAllocator uses AllocatorControl; the PooledAllocatorControl is now just a reference to the parent allocator.

Result:
More general internal APIs and simpler implementation.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/api/MemoryManager.java [hunk 1]
```diff
@@ -159,7 +160,7 @@ static Buffer unsafeWrap(byte[] array) {
      * @return A {@link Buffer} instance with the given configuration.
      * @throws IllegalArgumentException For unknown {@link AllocationType}s.
      */
-    Buffer allocateShared(AllocatorControl allocatorControl, long size, Drop<Buffer> drop,
+    Buffer allocateShared(AllocatorControl allocatorControl, long size, Function<Drop<Buffer>, Drop<Buffer>> drop,
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/MemoryManager.java:163]:
```suggestion
    Buffer allocateShared(AllocatorControl allocatorControl, long size, Function<Drop<Buffer>, Drop<Buffer>> dropDecorator,
```

File: buffer/src/main/java/io/netty/buffer/api/internal/DropCaptor.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.util.function.Function;
+
+/**
+ * Utility class to capture or hold on to a drop instance.
+ *
+ * @param <T> The type argument to {@link Drop}.
+ */
+public class DropCaptor<T> implements Function<Drop<T>, Drop<T>> {
+    public Drop<T> drop;
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/internal/DropCaptor.java:28]: Lets add a getter here and change this to private. Expose fields as public is a code-smell imho

File: buffer/src/main/java/io/netty/buffer/api/internal/DropCaptor.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import io.netty.buffer.api.Drop;
+
+import java.util.function.Function;
+
+/**
+ * Utility class to capture or hold on to a drop instance.
+ *
+ * @param <T> The type argument to {@link Drop}.
+ */
+public class DropCaptor<T> implements Function<Drop<T>, Drop<T>> {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/internal/DropCaptor.java:27]:
```suggestion
public final class DropCaptor<T> implements Function<Drop<T>, Drop<T>> {
```

File: buffer/src/main/java/io/netty/buffer/api/MemoryManager.java (hunk 1 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/MemoryManager.java:163]: Also update the javadocs

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Just some small things... I like it

**Review by normanmaurer — APPROVED** [review]

**Review by normanmaurer — APPROVED** [review]

**Comment by chrisvest** [comment]
Once this is merged, I'll push a matching update to https://github.com/netty/netty-incubator-buffer-api/pull/77

**Pull request merged**